### ISSUE TITLE
Debate: Phantom / Unused keyword

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#2650]
What about:

```scala
pub enum Vector[phantom imm: Bool, a: Type] {
    case Vector(Array[a])
}
```

and/or

```scala
unused def helloWorld(): String = ...
```

As an alternative (or addition) to using `_`.

Yes/no/maybe?

**Comment by mlutze** [comment]
`phantom` seems useful for communicating intent

I'm less convinced of `unused`. In general, we don't allow unused code. Surely adding a feature for it weakens that principle?

**Comment by magnus-madsen** [comment]
So what's the principle? Type variables can be marked as phantom?

Question: We could also just consider Boolean type variables as always used. (Or if you will, we only require that type variables of kind Type are actually used).

**Comment by JonathanStarup** [comment]
what is the argument for this? is it for wip convenience?

**Comment by magnus-madsen** [comment]
Some type variables are never used.. On purpose (see e.g. Vector issue). Thus it would seem strange to force you to write `_` when you are  actually using it in the way you intend.

**Comment by magnus-madsen** [comment]
A phantom (type variable) is a type variable that is used to enforce constraints on the code, but not for actual values.

**Comment by JonathanStarup** [comment]
ah okay, that makes sense. and the definition of "used" cant be extended to cover this kind of use?

**Comment by magnus-madsen** [comment]
> ah okay, that makes sense. and the definition of "used" cant be extended to cover this kind of use?

Well, it is *unused* in our current definition. We could say that since it has kind non-Type it does not have to be used.

**Comment by magnus-madsen** [comment]
Unless there are objections, I suggest that we disable the redundancy check for non-Star type variables.

**Comment by mlutze** [comment]
I object!
1. Suppose we want a phantom variable with 3 values. Under this rule we either have to mark it unused `_` (unfortunate), or create our phantom types as `Type -> Type` enums (very unfortunate), or do something weird with rows or something (very, very unfortunate).
2. Suppose I forget to use my `Type -> Type` variable in my enum. I think that's more likely a bug than a phantom variable.

In conclusion, my order of preference is:
1. Some keyword support
2. No special support
3. Support only for booleans

**Comment by magnus-madsen** [comment]
I am not sure what you mean by a phantom variable with 3 values?

**Comment by mlutze** [comment]
For example,
```scala

enum Red
enum Blue
enum Green

enum StringWithPhantomColor[_c] {
  case StringWithPhantomColor(String)
}

pub def f(): StringWithPhantomColor[Red] = ???
```

**Comment by JonathanStarup** [comment]
> For example,
> 
> ```scala
> enum Red
> enum Blue
> enum Green
> 
> enum StringWithPhantomColor[_c] {
>   case StringWithPhantomColor(String)
> }
> 
> pub def f(): StringWithPhantomColor[Red] = ???
> ```

I think `phantom` here could be nice. you mark that you use the type for modelling/constraints and the compiler knows that it shouldn't have values and that it can be erased after type checking

**Comment by magnus-madsen** [comment]
We currently have very few use cases for phantom types. If that should change, we can revisit. Until then it seems reasonable that they simply follow the same style as unused type variables (since that is what they are).
